sympy/matrices/matrices.py

```diff
@@ -1594,6 +1596,12 @@ def __pow__(self, num):
                     return self._matrix_pow_by_recursion(num)
             return self._matrix_pow_by_recursion(num)
         elif isinstance(num, (Expr, float)):
+            num = sympify(num)
+            if (num >= 0) == True and not num % 1 == 0:
+                z = self.jordan_form()[1]
```

Comment:
The only thing that I would consider changing in `__pow__` is sympifiying `num`. If that is done, it is not necessary to ask if `num` is an `int` or  a `float`. Instead, the tests could be `num.is_Integer` and `isinstance(num, Expr)`. Also, the tests in `_matrix_pow_by_jordan_blocks` could rely on `num` being sympified. But maybe that is not important and not necessary now.

As to the conditions in `jordan_cell_power`, I would take them in different order, first if `l == 0` and `jc.rows > 1`, and only then ask if `n` is a non-negative integer or not. That way, only one condition (`l == 0`) would have to be asked in most cases.